### PR TITLE
refactor(sbb-calendar): implement initial support for other date libraries

### PR DESCRIPTION
**Pull request by kyubisation**
This change allows us in the future to support different date libraries. However, this is only the implementation for the calendar. Datepicker will be done in a different PR.

BREAKING CHANGE:
The `SbbDatepicker` property `selectedDate` has been renamed to `selected`. This also applies to the attribute `selected-date`, which has been renamed to `selected`. Additionally the `DateAdapter` (and `NativeDateAdapter`) have been superficially refactored. An important change is that the month is now `1`-based, instead of `0`-based.

### CODE REVIEW

File: src/components/calendar/calendar.ts [hunk 1]
```diff
@@ -809,9 +817,9 @@ export class SbbCalendarElement extends LitElement {
         <div class="sbb-calendar__controls-month">
           ${this._createLabelForDayView(this._activeDate)}
           ${this._wide ? this._createLabelForDayView(nextMonthActiveDate!) : nothing}
-          <sbb-screenreader-only role="status">
+          <span role="status" class="sbb-calendar__visually-hidden">
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:820]: Could this be a rebase mistake? I recently changed to use sbb-screenreader-only

File: src/components/calendar/calendar.ts [hunk 2]
```diff
@@ -889,7 +897,7 @@ export class SbbCalendarElement extends LitElement {
     return this._weekdays.map(
       (day: Weekday) => html`
         <th class="sbb-calendar__table-header">
-          <sbb-screenreader-only>${day.long}</sbb-screenreader-only>
+          <span class="sbb-calendar__visually-hidden">${day.long}</span>
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:900]:
```suggestion
          <sbb-screenreader-only>${day.long}</sbb-screenreader-only>
```

File: src/components/calendar/calendar.ts [hunk 3]
```diff
@@ -1001,7 +1007,7 @@ export class SbbCalendarElement extends LitElement {
         ${this._chosenYear} ${this._wide ? ` - ${this._chosenYear! + 1}` : nothing}
         <sbb-icon name="chevron-small-up-small"></sbb-icon>
       </button>
-      <sbb-screenreader-only role="status"> ${this._chosenYear} </sbb-screenreader-only>`;
+      <span role="status" class="sbb-calendar__visually-hidden"> ${this._chosenYear} </span>`;
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:1010]:
```suggestion
      <sbb-screenreader-only role="status"> ${this._chosenYear} </sbb-screenreader-only>`;
```

Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:1010]:
```suggestion
      <sbb-screenreader-only role="status"> ${this._chosenYear} </sbb-screenreader-only>`;
```

File: src/components/calendar/calendar.ts [hunk 4]
```diff
@@ -1142,12 +1154,13 @@ export class SbbCalendarElement extends LitElement {
         ${yearLabel}
         <sbb-icon name="chevron-small-up-small"></sbb-icon>
       </button>
-      <sbb-screenreader-only role="status"> ${yearLabel} </sbb-screenreader-only>
+      <span role="status" class="sbb-calendar__visually-hidden"> ${yearLabel} </span>
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:1157]:
```suggestion
      <sbb-screenreader-only role="status"> ${yearLabel} </sbb-screenreader-only>
```

File: src/components/calendar/calendar.ts [hunk 5]
```diff
@@ -26,12 +26,10 @@ import {
   i18nYearMonthSelection,
 } from '../core/i18n';
 import type { SbbDateLike } from '../core/interfaces';
-
-import style from './calendar.scss?lit&inline';
-
 import '../button/secondary-button';
 import '../icon';
-import '../screenreader-only';
+
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.ts:31]:
```suggestion

import '../screenreader-only';
```

File: src/components/core/datetime/date-adapter.ts [hunk 6]
```diff
@@ -1,117 +1,115 @@
...
-   * @param date The ISO String to convert to date.
-   * */
-  createDateFromISOString: (date: string) => T | null;
-
-  /**
-   * Creates a Date from a valid input.
+   * Attempts to deserialize a value to a valid date object. This is different from parsing in that
+   * deserialize should only accept non-ambiguous, locale-independent formats (e.g. a ISO 8601
+   * string). The default implementation does not allow any deserialization, it simply checks that
+   * the given value is already a valid date object or null. The `<mat-datepicker>` will call this
```
Comment (jeripeierSBB) [inline src/components/core/datetime/date-adapter.ts:90]: seems to be copied docs of material?

File: src/components/core/datetime/date-adapter.ts [hunk 7]
```diff
@@ -120,46 +118,119 @@ export interface DateAdapter<T = any> {
...
+  public abstract format(date: T | null | undefined): string;
 
   /** Checks whether the given `obj` is a Date.
    * @param obj The object to check.
    */
-  isDateInstance: (obj: any) => boolean;
+  public abstract isDateInstance(obj: any): boolean;
+
+  /**
+   * Gets date instance that is not valid.
```
Comment (jeripeierSBB) [inline src/components/core/datetime/date-adapter.ts:152]: hard to understand the sense for me.hmm

File: src/components/calendar/calendar.e2e.ts [hunk 8]
```diff
@@ -6,8 +6,9 @@ import { waitForCondition, waitForLitRender, EventSpy } from '../core/testing';
 
 import { SbbCalendarElement } from './calendar';
 
-describe('sbb-calendar', () => {
-  const selected = new Date(2023, 0, 15).getTime() / 1000;
+import '../button';
+
+describe(`sbb-calendar with ${fixture.name}`, () => {
```
Comment (jeripeierSBB) [inline src/components/calendar/calendar.e2e.ts:11]: probably doesn't matter, but this change will come with your next pr :)
```suggestion
describe(`sbb-calendar`, () => {
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
As it is fairly complex, it's difficult to determine wether everything is fully like expected :). But generally looks good.
As testing fails, there is so far no build to manually experiment with the changes.

File: src/components/core/datetime/date-adapter.ts (hunk 7 above)
Comment (kyubisation) [reply src/components/core/datetime/date-adapter.ts:152]: This is a method that generates an intentionally invalid date.

Comment (jeripeierSBB) [reply src/components/core/datetime/date-adapter.ts:149]: You should once explain me the sense ;). but surely ok to merge without knowing exactly :)

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Codewise ok, but see chromatic changes, something with disabling buttons doesn't work anymore? or not the same?

**Review by jeripeierSBB — APPROVED** [review]
LGTM. thanks for the effort!

**Review by MarioCastigliano — APPROVED** [review]
Love the getter/setter approach on the calendar 👍

**Pull request merged**